### ISSUE TITLE
Feature request for only_predict

### ISSUE THREAD

**Issue by adg-git** [facebookresearch/SEAL_OGB#13]
Please could you add only_predict feature (similar to what you have in SEAL repo). My goal is to generate prediction scores for a set of links (provided in a separate file as input) in the graph. Thanks.

**Comment by adg-git** [comment]
Hi I tried this code for inference after calling model.eval and Dataloader 

for data in tqdm(infer_loader, ncols=70):
             data = data.to(device)
             x = data.x if args.use_feature else None
             edge_weight = data.edge_weight if args.use_edge_weight else None
             node_id = data.node_id if emb else None
             with torch.no_grad():
                  logits = model(data.z, data.edge_index, data.batch, x, edge_weight, node_id)


I find that the number of predictions that are returned are of different length than number of edges passed. I was hoping if edge_index=[2, 60] then logits.view(-1).cpu() will also have 60 values but that is not the case. 
When data is -> Batch(batch=[94], edge_index=[2, 60], edge_weight=[60], node_id=[94], y=[32], z=[94])
I get predictions of size 32 but was expecting 60
Could you throw some light on what could be the issue here.

**Comment by muhanzhang** [comment]
The `data` input of SEAL is a batch of subgraphs. If you set batch size to 32, then there are 32 subgraphs and 32 graph labels (corresponding to 32 link existences) in `data`. There will be more than 32 edges in `data` because each subgraph can contain more than one edge. 

Batch(batch=[94], edge_index=[2, 60], edge_weight=[60], node_id=[94], y=[32], z=[94]) means there are totally 94 nodes in a batch of 32 graphs, where the total number of edges is 60. You can check the data format of pytorch_geometric for more details.

**Comment by adg-git** [comment]
Thanks for the explanation. I am trying to map the prediction score to the pair of nodes for which it is inferring the link prediction . Any thoughts on how to map the two. So if I look at one subgraph
Data(edge_index=[2, 4], edge_weight=[4], node_id=[4], y=[1], z=[4])

This has four nodes. Which two nodes are we predicting the link score for?

**Comment by muhanzhang** [comment]
Yest. The first two node_ids are the target link.

**Comment by adg-git** [comment]
For model inference only this model initialization call
model=DGCNN(args.hidden_channels, args.num_layers, max_z,args.sortpool_k,
train_dataset,args.dynamic_train, use_feature=args.use_feature,
node_embedding=None)

needs train_dataset? Once I run to train and save the model and later I run to predict only and at that point I don't have train_dataset stored

**Comment by muhanzhang** [comment]
I see. Actually DGCNN only uses train_dataset during initialization for reading the feature dimension and setting the SortPooling `k`. If you no longer have a train_dataset during testing, you can record the `k` and manually input this integer number to DGCNN. You can also manually add your feature_dimension to this [line](https://github.com/facebookresearch/SEAL_OGB/blob/db81b1bb648f83e164411b977e6f52452193e125/models.py#L165) if your SEAL model uses node features.

**Comment by muhanzhang** [comment]
Yes. The models were not written to restrict the output to be probability, because OGB adopts the ranking metric such as Hits and MRR, where only relative order between two links matters.

**Comment by bits-glitch** [comment]
@adg-git did you manage to implement the only_predict function? How did you construct your inference loader?

**Comment by adg-git** [comment]
@bits-glitch yes I did. Its a separate child  class of InMemoryDataset that you can import from torch_geometric.data

**Comment by bits-glitch** [comment]
So you basically implemented your own inference dataset class that inherits InMemoryDataset and converts the links of your input file into the Data (data.edge_index,...) format?

**Comment by adg-git** [comment]
@bits-glitch yes that is what I did.

**Comment by TheDataCoder** [comment]
@muhanzhang sorry for bumping this up! Thank you for your hard work on this repo. 

I have similar question about targets in each subgraph: is there only a single target per subgraph (first two nodes in node_id) or there could be multiple targets?

**Comment by muhanzhang** [comment]
@TheDataCoder Hi! There is only a single target per subgraph (the first two nodes in node_id).

**Comment by TheDataCoder** [comment]
> @TheDataCoder Hi! There is only a single target per subgraph (the first two nodes in node_id).

Thank you! What about directed graph? How does algorithm works with it? Does it create sub-graphs for each direction between two nodes?

**Comment by muhanzhang** [comment]
Hi! For directed links, you can additionally predict its direction after predicting its existence, and you don't need to create a subgraph for each direction because both directions create the same subgraph. For some more details, see #12.
